xapian-core/diversify/diversify.cc

```diff
@@ -186,23 +172,45 @@ Diversify::Internal::get_dmset(const MSet& mset)
 	k = mset.size();
 
     initialise_points(mset);
-    compute_similarities();
+
+    // Cluster the given mset into k clusters
+    KMeans km(k);
+    Xapian::ClusterSet cset = km.cluster(mset);
+    compute_similarities(cset);
+
+    // topC contains union of top-r relevant documents of each cluster
+    vector<Xapian::docid> topc;
+
+    // Build topC
+    for (unsigned int c = 0; c < cset.size(); ++c) {
+	auto documents = cset[c].get_documents();
+	for (unsigned int d = 0; d < r && d < documents.size(); ++d) {
+	    auto doc_id = documents[d].get_docid();
+	    topc.push_back(doc_id);
+	}
+    }
 
     vector<Xapian::docid> curr_dmset = main_dmset;
 
     while (true) {
 	bool found_better_dmset = false;
 	for (unsigned int i = 0; i < main_dmset.size(); ++i) {
 	    auto curr_doc = main_dmset[i];
-	    double best_score = evaluate_dmset(curr_dmset);
+	    double best_score = evaluate_dmset(curr_dmset, cset);
+	    bool found_better_doc = false;
 
-	    vector<Xapian::docid> diff_dmset = compute_diff_dmset(curr_dmset);
+	    for (unsigned int j = 0; j < topc.size(); ++j) {
+		// Continue if candidate document from topC already
+		// exists in curr_dmset
+		auto candidate_doc = find(curr_dmset.begin(), curr_dmset.end(),
+					  topc[j]);
+		if (candidate_doc != curr_dmset.end()) {
+		    continue;
+		}
 
-	    bool found_better_doc = false;
-	    for (unsigned int j = 0; j < diff_dmset.size(); ++j) {
-		vector<Xapian::docid> temp_dmset = curr_dmset;
-		temp_dmset[i] = diff_dmset[j];
-		double score = evaluate_dmset(temp_dmset);
+		auto temp_dmset = curr_dmset;
+		temp_dmset[i] = topc[j];
+		double score = evaluate_dmset(temp_dmset, cset);
```

Comment:
So if I follow the code correctly here we make a copy of the whole of `curr_dmset` and change one entry in the copy, test that, and then throw away the copy.

It would probably be more efficient to actually modify that single element in `curr_dmset`, test the modified `curr_dmest` and then undo the change (we'd just need to store the old value of `curr_dmset[i]` in a temporary variable to be able to undo the change).

Reply:
This definitely is more efficient, thanks for pointing this out. I've pushed a fixup commit to reflect this change.